src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -127,6 +164,55 @@ public void AddEntriesFromPreviousTable(NodeStateTable<T> previousTable, EntrySt
                 _states.Add(previousEntries);
             }
 
+            public void ModifyEntriesFromPreviousTable(NodeStateTable<T> previousTable, ImmutableArray<T> outputs)
+            {
+                // Semantics:
+                // For every slot in the previous table, we compare the new value.
+                // - Cached when the same
+                // - Modified when different
+                // - Removed when i > outputs.length
+                // - Added when i < previousTable.length
+
+                var previousEntries = previousTable._states[_states.Count];
+                var modifiedEntries = ArrayBuilder<(T item, EntryState state)>.GetInstance();
+
+                var previousEnumerator = previousEntries.GetEnumerator();
+                var outputEnumerator = outputs.GetEnumerator();
+
+                bool previousHasItems = previousEnumerator.MoveNext();
+                bool outputHasItems = outputEnumerator.MoveNext();
+
+                // cached or modified items
+                while (previousHasItems && outputHasItems)
+                {
+                    var previous = previousEnumerator.Current;
+                    var replacement = outputEnumerator.Current;
+
+                    // PROTOTYPE(source-generators): support custom IEqualityComparer
```

Comment:
Ah this may answer one of my other questions about ordering. If the generator produces differently order'd output for same input it will be caught here. It will end up producing different entries in the table (two different orders could end up changing what come through as `Add` and what come through  as `Modified`). That should still *function* it just wouldn't be as cache friendly as stable ordered outputs.

Reply:
Bingo.

---
In reply to: [619519422](https://github.com/dotnet/roslyn/pull/52809#discussion_r619519422) [](ancestors = 619519422)